controller/controller.go

```diff
@@ -241,8 +256,28 @@ func (c *Controller) Reconcile(req reconcile.Request) (reconcile.Result, error)
 		ctx = setLoggerCtxValue(ctx, loggerKeyController, c.name)
 	}
 
-	res, err := c.reconcile(ctx, req)
+	obj := c.newRuntimeObjectFunc()
+	err := c.k8sClient.CtrlClient().Get(ctx, req.NamespacedName, obj)
+	if errors.IsNotFound(err) {
+		// At this point the controller-runtime cache dispatches a runtime object
+		// which is already being deleted, which is why it cannot be found here
+		// anymore. We then likely perceive the last delete event of that runtime
+		// object and it got purged from the controller-runtime cache. We do not
+		// need to log these errors and just stop processing here in a more graceful
+		// way.
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		return reconcile.Result{}, microerror.Mask(err)
+	}
+
+	res, err := c.reconcile(ctx, req, obj)
 	if err != nil {
+		// In case of microerror create an event on the object
+		if merr, ok := microerror.Cause(err).(*microerror.Error); ok {
+			if merr.Kind != "" && merr.Desc != "" {
+				c.recorder.Event(obj, corev1.EventTypeWarning, merr.Kind, merr.Desc)
+			}
+		}
```

Comment:
Bonus points for making it pretty. 

```suggestion
		c.event.Emit(ctx, obj, err)
```

Reply:
I like that 👍